File: android/src/main/java/expo/modules/applemusic/ExpoAppleMusicModule.kt
package expo.modules.applemusic

import com.apple.android.music.playback.model.MediaPlayerException
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.activityresult.AppContextActivityResultLauncher
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

class ExpoAppleMusicModule : Module() {
  private lateinit var authLauncher: AppContextActivityResultLauncher<MusicKitAuthInput, MusicKitAuthOutput>
  private var playbackObserver: AndroidPlaybackObserver? = null

  private val moduleScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

  private val reactContext
    get() = requireNotNull(appContext.reactContext) { "React Application Context is null" }

  private val playbackController: AndroidPlaybackController
    get() =
      AndroidPlaybackController.getInstance(reactContext).also { controller ->
        if (controller.playbackErrorHandler == null) {
          controller.playbackErrorHandler = { error, operation ->
            emitPlaybackError(error, operation)
          }
        }
      }

  private val catalogService: AndroidCatalogService
    get() = AndroidCatalogService(reactContext)

  private val libraryService: AndroidLibraryService
    get() = AndroidLibraryService(reactContext)

  private val subscriptionService: AndroidSubscriptionService
    get() = AndroidSubscriptionService(reactContext)

  private val queueService: AndroidQueueService
    get() = AndroidQueueService(reactContext, playbackController)

  private val ratingsService: AndroidRatingsService
    get() = AndroidRatingsService(reactContext)

  private val libraryMutationsService: AndroidLibraryMutationsService
    get() = AndroidLibraryMutationsService(reactContext)

  override fun definition() = ModuleDefinition {
    Name("ExpoAppleMusic")

    Events(
      "onPlaybackStateChange",
      "onCurrentSongChange",
      "onPlaybackTimeUpdate",
      "onPlaybackError",
    )

    OnStartObserving {
      val observer = AndroidPlaybackObserver(playbackController)
      observer.delegate =
        object : AndroidPlaybackObserverDelegate {
          override fun onPlaybackStateChange(body: Map<String, Any?>) {
            sendEvent("onPlaybackStateChange", body)
          }

          override fun onCurrentSongChange(body: Map<String, Any?>) {
            sendEvent("onCurrentSongChange", body)
          }

          override fun onPlaybackTimeUpdate(playbackTime: Double) {
            sendEvent("onPlaybackTimeUpdate", mapOf("playbackTime" to playbackTime))
          }

          override fun onPlaybackError(body: Map<String, Any?>) {
            sendEvent("onPlaybackError", body)
          }
        }
      observer.startObserving()
      playbackObserver = observer
    }

    OnStopObserving {
      playbackObserver?.stopObserving()
      playbackObserver = null
    }

    RegisterActivityContracts {
      authLauncher =
        registerForActivityResult(
          MusicKitAuthContract {
            appContext.throwingActivity
          },
        )
    }

    AsyncFunction("authorization") Coroutine {
      developerToken: String?,
      startScreenMessage: String?,
      hideStartScreen: Boolean?,
    ->
      val context = reactContext
      val token = AndroidDeveloperToken.require(developerToken)
      MusicKitAuthStorage.saveDeveloperToken(context, token)
      val message = startScreenMessage?.trim()?.takeIf { it.isNotEmpty() }
      val result =
        authLauncher.launch(
          MusicKitAuthInput(
            developerToken = token,
            startScreenMessage = message,
            hideStartScreen = hideStartScreen ?: false,
          ),
        )

      result.musicUserToken?.let { MusicKitAuthStorage.saveMusicUserToken(context, it) }
      if (result.status == "authorized") {
        AndroidPlaybackController.warmUp(context)
      }
      result.status
    }

    AsyncFunction("checkSubscription") Coroutine { ->
      subscriptionService.checkSubscription()
    }

    AsyncFunction("catalogSearch") Coroutine {
      term: String,
      types: List<String>,
      options: Map<String, Any?>,
    ->
      val pagination = PaginationOptions.fromMap(options)
      val result = catalogService.search(term, types, pagination)
      mapOf(
        "songs" to result.songs,
        "albums" to result.albums,
        "artists" to result.artists,
        "playlists" to result.playlists,
        "stations" to result.stations,
        "musicVideos" to result.musicVideos,
      )
    }

    AsyncFunction("getCatalogSong") Coroutine { id: String ->
      catalogService.getSong(id)
    }

    AsyncFunction("getCatalogAlbum") Coroutine { id: String ->
      catalogService.getAlbum(id)
    }

    AsyncFunction("getCatalogArtist") Coroutine { id: String ->
      catalogService.getArtist(id)
    }

    AsyncFunction("getCatalogPlaylist") Coroutine { id: String ->
      catalogService.getPlaylist(id)
    }

    AsyncFunction("getCatalogStation") Coroutine { id: String ->
      catalogService.getStation(id)
    }

    AsyncFunction("getCatalogMusicVideo") Coroutine { id: String ->
      catalogService.getMusicVideo(id)
    }

    AsyncFunction("getCatalogAlbumTracks") Coroutine { albumId: String, options: Map<String, Any?> ->
      val pagination = PaginationOptions.fromMap(options)
      val songs = catalogService.getAlbumTracks(albumId, pagination)
      mapOf("songs" to songs)
    }

    AsyncFunction("getCatalogArtistAlbums") Coroutine { artistId: String, options: Map<String, Any?> ->
      val pagination = PaginationOptions.fromMap(options)
      val albums = catalogService.getArtistAlbums(artistId, pagination)
      mapOf("albums" to albums)
    }

    AsyncFunction("getCatalogPlaylistTracks") Coroutine { playlistId: String, options: Map<String, Any?> ->
      val pagination = PaginationOptions.fromMap(options)
      val songs = catalogService.getPlaylistTracks(playlistId, pagination)
      mapOf("songs" to songs)
    }

    AsyncFunction("getCatalogCharts") Coroutine { types: List<String>, options: Map<String, Any?> ->
      val pagination = PaginationOptions.fromMap(options)
      val genre = options["genre"] as? String
      val chart = options["chart"] as? String
      val result = catalogService.getCharts(types, pagination, genre, chart)
      mapOf(
        "songs" to result.songs,
        "albums" to result.albums,
        "playlists" to result.playlists,
        "musicVideos" to result.musicVideos,
      )
    }

    AsyncFunction("setPlaybackQueue") Coroutine { itemId: String, type: String ->
      try {
        queueService.setQueue(itemId, type)
        "Track(s) added to queue"
      } catch (error: Exception) {
        throw AndroidPlaybackController.mapPlaybackException(error)
      }
    }

    AsyncFunction("getStorefront") Coroutine { ->
      mapOf("id" to libraryService.getStorefrontId())
    }

    AsyncFunction("getTracksFromLibrary") Coroutine { ->
      val tracks = libraryService.getRecentlyPlayed()
      mapOf("recentlyPlayedItems" to tracks)
    }

    AsyncFunction("getRecentlyPlayedTracks") Coroutine { options: Map<String, Any?> ->
      val pagination = PaginationOptions.fromMap(options)
      val songs = libraryService.getRecentlyPlayedTracks(pagination)
      mapOf("songs" to songs)
    }

    AsyncFunction("getLibraryArtists") Coroutine { options: Map<String, Any?> ->
      val pagination = PaginationOptions.fromMap(options)
      val artists = libraryService.getArtists(pagination)
      mapOf("artists" to artists)
    }

    AsyncFunction("getLibraryAlbums") Coroutine { options: Map<String, Any?> ->
      val pagination = PaginationOptions.fromMap(options)
      val albums = libraryService.getAlbums(pagination)
      mapOf("albums" to albums)
    }

    AsyncFunction("getHeavyRotation") Coroutine { options: Map<String, Any?> ->
      val pagination = PaginationOptions.fromMap(options)
      val items = libraryService.getHeavyRotation(pagination)
      mapOf("items" to items)
    }

    AsyncFunction("getRecentlyPlayedStations") Coroutine { options: Map<String, Any?> ->
      val pagination = PaginationOptions.fromMap(options)
      val stations = libraryService.getRecentlyPlayedStations(pagination)
      mapOf("stations" to stations)
    }

    AsyncFunction("getRecentlyAdded") Coroutine { options: Map<String, Any?> ->
      val pagination = PaginationOptions.fromMap(options)
      val items = libraryService.getRecentlyAdded(pagination)
      mapOf("items" to items)
    }

    AsyncFunction("configurePlayer") { mixWithOthers: Boolean ->
      playbackController.configurePlayer(mixWithOthers)
    }

    AsyncFunction("getCurrentState") {
      playbackController.currentState()
    }

    AsyncFunction("getUserPlaylists") Coroutine { options: Map<String, Any?> ->
      val pagination = PaginationOptions.fromMap(options)
      val playlists = libraryService.getPlaylists(pagination)
      mapOf("playlists" to playlists)
    }

    AsyncFunction("getLibrarySongs") Coroutine { options: Map<String, Any?> ->
      val pagination = PaginationOptions.fromMap(options)
      val songs = libraryService.getSongs(pagination)
      mapOf("songs" to songs)
    }

    AsyncFunction("getPlaylistSongs") Coroutine { playlistId: String, options: Map<String, Any?> ->
      val songs = libraryService.getPlaylistSongs(playlistId)
      mapOf("songs" to songs)
    }

    AsyncFunction("playLibrarySong") Coroutine { songId: String ->
      queueService.playLibrarySong(songId)
      "Library song added to queue"
    }

    AsyncFunction("playLibraryPlaylist") Coroutine { playlistId: String, startingAt: Int ->
      queueService.playLibraryPlaylist(playlistId, startingAt)
      "Library playlist added to queue"
    }

    AsyncFunction("getRating") Coroutine { resourceType: String, id: String ->
      ratingsService.getRating(resourceType, id)
    }

    AsyncFunction("setRating") Coroutine { resourceType: String, id: String, value: Int ->
      ratingsService.setRating(resourceType, id, value)
    }

    AsyncFunction("clearRating") Coroutine { resourceType: String, id: String ->
      ratingsService.clearRating(resourceType, id)
    }

    AsyncFunction("addToFavorites") Coroutine { resourceIds: Map<String, List<String>> ->
      ratingsService.addToFavorites(resourceIds)
    }

    AsyncFunction("removeFromFavorites") Coroutine { resourceIds: Map<String, List<String>> ->
      ratingsService.removeFromFavorites(resourceIds)
    }

    AsyncFunction("addToLibrary") Coroutine { resourceIds: Map<String, List<String>> ->
      libraryMutationsService.addToLibrary(resourceIds)
    }

    AsyncFunction("createLibraryPlaylist") Coroutine { options: Map<String, Any?> ->
      val name = options["name"] as? String ?: ""
      val description = options["description"] as? String
      val isPublic = options["isPublic"] as? Boolean ?: false
      @Suppress("UNCHECKED_CAST")
      val tracks = options["tracks"] as? List<Map<String, String>>
      libraryMutationsService.createPlaylist(name, description, isPublic, tracks)
    }

    AsyncFunction("addTracksToLibraryPlaylist") Coroutine {
      playlistId: String,
      tracks: List<Map<String, String>>,
    ->
      libraryMutationsService.addTracksToPlaylist(playlistId, tracks)
    }

    Function("play") {
      runPlayback("play") { playbackController.play() }
    }

    Function("pause") {
      playbackController.pause()
    }

    Function("skipToNextEntry") {
      runPlayback("skipToNext") { playbackController.skipToNext() }
    }

    Function("skipToPreviousEntry") {
      runPlayback("skipToPrevious") { playbackController.skipToPrevious() }
    }

    Function("restartCurrentEntry") {
      playbackController.restartCurrentEntry { time ->
        sendEvent("onPlaybackTimeUpdate", mapOf("playbackTime" to time))
      }
    }

    Function("seekToTime") { time: Double ->
      playbackController.seekToTime(time) { actual ->
        sendEvent("onPlaybackTimeUpdate", mapOf("playbackTime" to actual))
      }
    }

    Function("togglePlayerState") {
      runPlayback("togglePlayback") { playbackController.togglePlayback() }
    }
  }

  private fun runPlayback(operation: String, block: () -> Unit) {
    moduleScope.launch {
      try {
        block()
      } catch (error: MediaPlayerException) {
        emitPlaybackError(error, operation)
      } catch (error: Exception) {
        emitPlaybackError(error, operation)
      }
    }
  }

  private fun emitPlaybackError(error: Exception, operation: String) {
    val code =
      when (error) {
        is MediaPlayerException -> error.errorCode
        is CodedException -> error.code
        else -> 0
      }
    sendEvent(
      "onPlaybackError",
      mapOf(
        "message" to (error.message ?: "Playback error"),
        "code" to code,
        "domain" to error.javaClass.simpleName,
        "operation" to operation,
      ),
    )
  }
}

File: ios/ExpoAppleMusicModule.swift
import ExpoModulesCore

@available(iOS 16.0, *)
public class ExpoAppleMusicModule: Module {
  private let playbackController = PlaybackController.shared
  private let subscriptionService = SubscriptionService()
  private let catalogService = CatalogService()
  private lazy var queueService = QueueService(
    playbackController: playbackController,
    catalogService: catalogService
  )
  private lazy var libraryService = LibraryService()
  private lazy var historyService = HistoryService()
  private lazy var ratingsService = RatingsService()
  private lazy var libraryMutationsService = LibraryMutationsService()
  private lazy var recommendationsService = RecommendationsService()

  private var playbackObserver: PlaybackObserver?

  public func definition() -> ModuleDefinition {
    Name("ExpoAppleMusic")

      Events(
        "onPlaybackStateChange",
        "onCurrentSongChange",
        "onPlaybackTimeUpdate",
        "onPlaybackError"
      )

      OnStartObserving {
        let observer = PlaybackObserver(playbackController: self.playbackController)
        observer.delegate = self
        observer.startObserving()
        self.playbackObserver = observer
      }

      OnStopObserving {
        self.playbackObserver?.stopObserving()
        self.playbackObserver = nil
      }

      // MARK: - Auth

      AsyncFunction("setDeveloperToken") { (token: String) in
        MusicKitAuthStorage.saveDeveloperToken(token)
      }

      AsyncFunction("authorization") { (developerToken: String?, _ startScreenMessage: String?, _ hideStartScreen: Bool?) -> [String: Any?] in
        if let token = developerToken, !token.isEmpty {
          MusicKitAuthStorage.saveDeveloperToken(token)
        }
        let status = await self.subscriptionService.requestAuthorization()
        var musicUserToken: String? = nil
        if status == .authorized, let token = developerToken, !token.isEmpty {
          musicUserToken = await self.subscriptionService.fetchMusicUserToken(developerToken: token)
        }
        return ["status": status.rawValue, "musicUserToken": musicUserToken]
      }

      AsyncFunction("checkSubscription") { (musicUserToken: String) -> [String: Any] in
        do {
          let details = try await self.subscriptionService.checkSubscription()
          return details.toDictionary()
        } catch {
          if let subError = SubscriptionService.wrapSubscriptionError(error) {
            throw Exception(
              name: subError.code,
              description: subError.message,
              code: subError.code
            )
          }
          throw AppleMusicBridgeError.exception(from: error)
        }
      }

      AsyncFunction("getStorefront") { (musicUserToken: String) -> [String: Any] in
        try await AppleMusicBridgeError.rethrow {
          let id = try await StorefrontService.getStorefrontId(musicUserToken: musicUserToken)
          return BridgeResponses.storefront(id: id)
        }
      }

      // MARK: - Catalog

      AsyncFunction("catalogSearch") { (term: String, types: [String], options: [String: Any]) -> [String: Any] in
        try await ExpoBridgeCatalog.catalogSearch(
          service: self.catalogService,
          term: term,
          types: types,
          options: options as NSDictionary
        )
      }

      AsyncFunction("getCatalogSong") { (id: String) -> [String: Any] in
        try await AppleMusicBridgeError.rethrow {
          try await self.catalogService.getSong(id: id)
        }
      }

      AsyncFunction("getCatalogAlbum") { (id: String) -> [String: Any] in
        try await AppleMusicBridgeError.rethrow {
          try await self.catalogService.getAlbum(id: id)
        }
      }

      AsyncFunction("getCatalogArtist") { (id: String) -> [String: Any] in
        try await AppleMusicBridgeError.rethrow {
          try await self.catalogService.getArtist(id: id)
        }
      }

      AsyncFunction("getCatalogPlaylist") { (id: String) -> [String: Any] in
        try await AppleMusicBridgeError.rethrow {
          try await self.catalogService.getPlaylist(id: id)
        }
      }

      AsyncFunction("getCatalogStation") { (id: String) -> [String: Any] in
        try await AppleMusicBridgeError.rethrow {
          try await self.catalogService.getStation(id: id)
        }
      }

      AsyncFunction("getCatalogMusicVideo") { (id: String) -> [String: Any] in
        try await AppleMusicBridgeError.rethrow {
          try await self.catalogService.getMusicVideo(id: id)
        }
      }

      AsyncFunction("getCatalogAlbumTracks") { (albumId: String, options: [String: Any]) -> [String: Any] in
        try await ExpoBridgeCatalog.getCatalogAlbumTracks(
          service: self.catalogService,
          albumId: albumId,
          options: options as NSDictionary
        )
      }

      AsyncFunction("getCatalogArtistAlbums") { (artistId: String, options: [String: Any]) -> [String: Any] in
        try await ExpoBridgeCatalog.getCatalogArtistAlbums(
          service: self.catalogService,
          artistId: artistId,
          options: options as NSDictionary
        )
      }

      AsyncFunction("getCatalogPlaylistTracks") { (playlistId: String, options: [String: Any]) -> [String: Any] in
        try await ExpoBridgeCatalog.getCatalogPlaylistTracks(
          service: self.catalogService,
          playlistId: playlistId,
          options: options as NSDictionary
        )
      }

      AsyncFunction("getCatalogCharts") { (types: [String], options: [String: Any]) -> [String: Any] in
        try await ExpoBridgeCatalog.getCatalogCharts(
          service: self.catalogService,
          types: types,
          options: options
        )
      }

      AsyncFunction("getCatalogResources") { (type: String, ids: [String]) -> [String: Any] in
        try await ExpoBridgeCatalog.getCatalogResources(
          service: self.catalogService,
          type: type,
          ids: ids
        )
      }

      // MARK: - Library

      AsyncFunction("getUserPlaylists") { (musicUserToken: String, options: [String: Any]) -> [String: Any] in
        try await ExpoBridgeLibrary.getUserPlaylists(
          service: self.libraryService,
          musicUserToken: musicUserToken,
          options: options as NSDictionary
        )
      }

      AsyncFunction("getLibrarySongs") { (musicUserToken: String, options: [String: Any]) -> [String: Any] in
        try await ExpoBridgeLibrary.getLibrarySongs(
          service: self.libraryService,
          musicUserToken: musicUserToken,
          options: options as NSDictionary
        )
      }

      AsyncFunction("getPlaylistSongs") { (musicUserToken: String, playlistId: String, options: [String: Any]) -> [String: Any] in
        try await ExpoBridgeLibrary.getPlaylistSongs(
          service: self.libraryService,
          musicUserToken: musicUserToken,
          playlistId: playlistId
        )
      }

      AsyncFunction("getLibraryArtists") { (musicUserToken: String, options: [String: Any]) -> [String: Any] in
        try await ExpoBridgeLibrary.getLibraryArtists(
          service: self.libraryService,
          musicUserToken: musicUserToken,
          options: options as NSDictionary
        )
      }

      AsyncFunction("getLibraryAlbums") { (musicUserToken: String, options: [String: Any]) -> [String: Any] in
        try await ExpoBridgeLibrary.getLibraryAlbums(
          service: self.libraryService,
          musicUserToken: musicUserToken,
          options: options as NSDictionary
        )
      }

      AsyncFunction("getLibraryMusicVideos") { (musicUserToken: String, options: [String: Any]) -> [String: Any] in
        try await ExpoBridgeLibrary.getLibraryMusicVideos(
          service: self.libraryService,
          musicUserToken: musicUserToken,
          options: options as NSDictionary
        )
      }

      AsyncFunction("librarySearch") { (musicUserToken: String, term: String, types: [String], options: [String: Any]) -> [String: Any] in
        try await ExpoBridgeLibrary.librarySearch(
          service: self.libraryService,
          musicUserToken: musicUserToken,
          term: term,
          types: types,
          options: options as NSDictionary
        )
      }

      // MARK: - History

      AsyncFunction("getRecentlyPlayedResources") { (musicUserToken: String) -> [String: Any] in
        try await ExpoBridgeHistory.getRecentlyPlayedResources(
          service: self.historyService,
          musicUserToken: musicUserToken)
      }

      AsyncFunction("getRecentlyPlayedTracks") { (musicUserToken: String, options: [String: Any]) -> [String: Any] in
        try await ExpoBridgeHistory.getRecentlyPlayedTracks(
          service: self.historyService,
          musicUserToken: musicUserToken,
          options: options as NSDictionary
        )
      }

      AsyncFunction("getHeavyRotation") { (musicUserToken: String, options: [String: Any]) -> [String: Any] in
        try await ExpoBridgeHistory.getHeavyRotation(
          service: self.historyService,
          musicUserToken: musicUserToken,
          options: options as NSDictionary
        )
      }

      AsyncFunction("getRecentlyPlayedStations") { (musicUserToken: String, options: [String: Any]) -> [String: Any] in
        try await ExpoBridgeHistory.getRecentlyPlayedStations(
          service: self.historyService,
          musicUserToken: musicUserToken,
          options: options as NSDictionary
        )
      }

      AsyncFunction("getRecentlyAdded") { (musicUserToken: String, options: [String: Any]) -> [String: Any] in
        try await ExpoBridgeHistory.getRecentlyAdded(
          service: self.historyService,
          musicUserToken: musicUserToken,
          options: options as NSDictionary
        )
      }

      // MARK: - Player

      AsyncFunction("setPlaybackQueue") { (itemId: String, type: String) -> String in
        try await AppleMusicBridgeError.rethrow {
          try await self.queueService.setQueue(itemId: itemId, type: type)
          return "Track(s) added to queue"
        }
      }

      AsyncFunction("configurePlayer") { (mixWithOthers: Bool) -> [String: Any] in
        try self.playbackController.configureAudioSession(mixWithOthers: mixWithOthers)
        return BridgeResponses.configurePlayer(mixWithOthers: mixWithOthers)
      }

      Function("play") {
        Task {
          do {
            try await self.playbackController.play()
          } catch {
            self.emitPlaybackError(error, operation: "play")
          }
        }
      }

      Function("pause") {
        self.playbackController.pause()
      }

      Function("skipToNextEntry") {
        Task {
          do {
            try await self.playbackController.skipToNext()
          } catch {
            self.emitPlaybackError(error, operation: "skipToNext")
          }
        }
      }

      Function("skipToPreviousEntry") {
        Task {
          do {
            try await self.playbackController.skipToPrevious()
          } catch {
            self.emitPlaybackError(error, operation: "skipToPrevious")
          }
        }
      }

      Function("restartCurrentEntry") {
        Task { @MainActor in
          self.playbackController.restartCurrentEntry()
          self.playbackTimeDidUpdate(0)
        }
      }

      Function("seekToTime") { (time: Double) in
        Task { @MainActor in
          self.playbackController.seek(to: time)
          self.playbackTimeDidUpdate(time)
        }
      }

      Function("togglePlayerState") {
        Task {
          do {
            try await self.playbackController.togglePlayback()
          } catch {
            self.emitPlaybackError(error, operation: "togglePlayback")
          }
        }
      }

      AsyncFunction("getCurrentState") { () -> [String: Any] in
        let state = self.playbackController.state
        let songInfo = await self.playbackController.fetchCurrentSongInfo()

        var result: [String: Any] = [
          "playbackRate": state.playbackRate,
          "playbackStatus": MusicItemMapper.describePlaybackStatus(state.playbackStatus),
          "playbackTime": self.playbackController.playbackTime,
        ]
        if let songInfo = songInfo {
          result["currentSong"] = songInfo
        }
        return result
      }

      AsyncFunction("playLibrarySong") { (musicUserToken: String, songId: String) -> String in
        try await AppleMusicBridgeError.rethrow {
          try await self.queueService.playLibrarySong(musicUserToken: musicUserToken, songId: songId)
          return "Library song added to queue"
        }
      }

      AsyncFunction("playLibraryPlaylist") { (musicUserToken: String, playlistId: String, startingAt: Int) -> String in
        try await AppleMusicBridgeError.rethrow {
          try await self.queueService.playLibraryPlaylist(
            musicUserToken: musicUserToken,
            playlistId: playlistId,
            startingAt: startingAt
          )
          return "Library playlist added to queue"
        }
      }

      // MARK: - Ratings

      AsyncFunction("getRating") { (musicUserToken: String, resourceType: String, id: String) -> [String: Any]? in
        try await AppleMusicBridgeError.rethrow {
          try await self.ratingsService.getRating(
            musicUserToken: musicUserToken, resourceType: resourceType, id: id)
        }
      }

      AsyncFunction("setRating") { (musicUserToken: String, resourceType: String, id: String, value: Int) -> [String: Any] in
        try await AppleMusicBridgeError.rethrow {
          try await self.ratingsService.setRating(
            musicUserToken: musicUserToken, resourceType: resourceType, id: id, value: value)
        }
      }

      AsyncFunction("clearRating") { (musicUserToken: String, resourceType: String, id: String) -> Void in
        try await AppleMusicBridgeError.rethrow {
          try await self.ratingsService.clearRating(
            musicUserToken: musicUserToken, resourceType: resourceType, id: id)
        }
      }

      AsyncFunction("addToFavorites") { (musicUserToken: String, resourceIds: [String: [String]]) -> Void in
        try await AppleMusicBridgeError.rethrow {
          try await self.ratingsService.addToFavorites(
            musicUserToken: musicUserToken, resourceIds: resourceIds)
        }
      }

      AsyncFunction("removeFromFavorites") { (musicUserToken: String, resourceIds: [String: [String]]) -> Void in
        try await AppleMusicBridgeError.rethrow {
          try await self.ratingsService.removeFromFavorites(
            musicUserToken: musicUserToken, resourceIds: resourceIds)
        }
      }

      // MARK: - Library mutations

      AsyncFunction("addToLibrary") { (musicUserToken: String, resourceIds: [String: [String]]) -> Void in
        try await AppleMusicBridgeError.rethrow {
          try await self.libraryMutationsService.addToLibrary(
            musicUserToken: musicUserToken, resourceIds: resourceIds)
        }
      }

      AsyncFunction("createLibraryPlaylist") { (musicUserToken: String, options: [String: Any]) -> [String: Any] in
        try await AppleMusicBridgeError.rethrow {
          let name = options["name"] as? String ?? ""
          let description = options["description"] as? String
          let isPublic = options["isPublic"] as? Bool ?? false
          let tracks = options["tracks"] as? [[String: String]]
          return try await self.libraryMutationsService.createPlaylist(
            musicUserToken: musicUserToken,
            name: name,
            description: description,
            isPublic: isPublic,
            tracks: tracks
          )
        }
      }

      AsyncFunction("addTracksToLibraryPlaylist") { (musicUserToken: String, playlistId: String, tracks: [[String: String]]) -> Void in
        try await AppleMusicBridgeError.rethrow {
          try await self.libraryMutationsService.addTracksToPlaylist(
            musicUserToken: musicUserToken,
            playlistId: playlistId,
            tracks: tracks
          )
        }
      }

      // MARK: - Recommendations

      AsyncFunction("getRecommendations") { (musicUserToken: String, ids: [String]?) -> [String: Any] in
        try await ExpoBridgeRecommendations.getRecommendations(
          service: self.recommendationsService,
          musicUserToken: musicUserToken,
          ids: ids
        )
      }

      AsyncFunction("getReplay") { (musicUserToken: String, year: Int?) -> [String: Any] in
        try await ExpoBridgeRecommendations.getReplay(
          service: self.recommendationsService,
          musicUserToken: musicUserToken,
          year: year
        )
      }
  }

  private func emitPlaybackError(_ error: Error, operation: String) {
    let nsError = error as NSError
    sendEvent(
      "onPlaybackError",
      [
        "message": error.localizedDescription,
        "code": nsError.code,
        "domain": nsError.domain,
        "operation": operation,
      ]
    )
  }
}

@available(iOS 16.0, *)
extension ExpoAppleMusicModule: PlaybackObserverDelegate {
  @MainActor
  func playbackStateDidChange(_ state: PlaybackObserver.PlaybackInfo) {
    var body: [String: Any] = [
      "playbackRate": state.playbackRate,
      "playbackStatus": state.playbackStatus,
      "playbackTime": state.playbackTime,
    ]
    if let song = state.currentSong {
      body["currentSong"] = song
    }
    sendEvent("onPlaybackStateChange", body)
  }

  @MainActor
  func currentSongDidChange(_ songInfo: [String: Any]?) {
    guard let songInfo = songInfo else { return }
    sendEvent("onCurrentSongChange", ["currentSong": songInfo])
  }

  @MainActor
  func playbackTimeDidUpdate(_ time: TimeInterval) {
    sendEvent("onPlaybackTimeUpdate", ["playbackTime": time])
  }
}
